UT failure: panic: duplicate metrics collector registration attempted

### Failure cluster [1b410c50dc9c60031af5](https://go.k8s.io/triage#1b410c50dc9c60031af5)
https://storage.googleapis.com/k8s-triage/index.html?text=prometheus

##### Error text:
**TestRecordOperation**
```
=== RUN   TestRecordOperation
--- FAIL: TestRecordOperation (0.00s)
panic: duplicate metrics collector registration attempted [recovered]
	panic: duplicate metrics collector registration attempted

goroutine 281 [running]:
testing.tRunner.func1.2({0x2b86420, 0xc000526a20})
	/usr/local/go/src/testing/testing.go:1209 +0x36c
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x3b6
panic({0x2b86420, 0xc000526a20})
	/usr/local/go/src/runtime/panic.go:1047 +0x266
k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc000164000, {0xc00081ac80, 0x1, 0x0})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus/registry.go:403 +0xa7
k8s.io/kubernetes/vendor/k8s.io/component-base/metrics.(*kubeRegistry).MustRegister(0xc000164000, {0xc00081ac70, 0x1, 0x10})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/component-base/metrics/registry.go:176 +0xcc
k8s.io/kubernetes/pkg/kubelet/kuberuntime.TestRecordOperation(0x0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/instrumented_services_test.go:33 +0xe4
testing.tRunner(0xc000602b60, 0x2e77168)
	/usr/local/go/src/testing/testing.go:1259 +0x230
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x727

```

**TestRecordOperation**
```
=== RUN   TestRecordOperation
--- FAIL: TestRecordOperation (0.00s)
panic: duplicate metrics collector registration attempted [recovered]
	panic: duplicate metrics collector registration attempted

goroutine 281 [running]:
testing.tRunner.func1.2({0x2b86420, 0xc000526a20})
	/usr/local/go/src/testing/testing.go:1209 +0x36c
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x3b6
panic({0x2b86420, 0xc000526a20})
	/usr/local/go/src/runtime/panic.go:1047 +0x266
k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc000164000, {0xc00081ac80, 0x1, 0x0})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus/registry.go:403 +0xa7
k8s.io/kubernetes/vendor/k8s.io/component-base/metrics.(*kubeRegistry).MustRegister(0xc000164000, {0xc00081ac70, 0x1, 0x10})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/component-base/metrics/registry.go:176 +0xcc
k8s.io/kubernetes/pkg/kubelet/kuberuntime.TestRecordOperation(0x0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/instrumented_services_test.go:33 +0xe4
testing.tRunner(0xc000602b60, 0x2e77168)
	/usr/local/go/src/testing/testing.go:1259 +0x230
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x727
```
**TestGetHistogramVecFromGatherer**

```
=== RUN   TestGetHistogramVecFromGatherer/filter_with_one_label
    --- FAIL: TestGetHistogramVecFromGatherer/filter_with_one_label (0.00s)
panic: duplicate metrics collector registration attempted [recovered]
	panic: duplicate metrics collector registration attempted

goroutine 131 [running]:
testing.tRunner.func1.2({0x9caea0, 0xc00008a7e0})
	/usr/local/go/src/testing/testing.go:1209 +0x36c
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x3b6
panic({0x9caea0, 0xc00008a7e0})
	/usr/local/go/src/runtime/panic.go:1047 +0x266
k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc000320000, {0xc000408b10, 0x1, 0x0})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus/registry.go:403 +0xa7
k8s.io/kubernetes/vendor/k8s.io/component-base/metrics.(*kubeRegistry).MustRegister(0xc000320000, {0xc000408b00, 0x1, 0x10})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/component-base/metrics/registry.go:176 +0xcc
k8s.io/kubernetes/vendor/k8s.io/component-base/metrics/testutil.TestGetHistogramVecFromGatherer.func1.1()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/component-base/metrics/testutil/metrics_test.go:580 +0xa2
sync.(*Once).doSlow(0xc00041ab84, 0xc000078e38)
	/usr/local/go/src/sync/once.go:68 +0x128
sync.(*Once).Do(0xc00041ab84, 0xc00008a760)
	/usr/local/go/src/sync/once.go:59 +0x47
k8s.io/kubernetes/vendor/k8s.io/component-base/metrics/testutil.TestGetHistogramVecFromGatherer.func1(0xc000419860)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/component-base/metrics/testutil/metrics_test.go:579 +0x35f
testing.tRunner(0xc000419860, 0xc0001415a8)
	/usr/local/go/src/testing/testing.go:1259 +0x230
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x727
```

#### Recent failures:
[2021/9/4 上午2:00:19 ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1433851876134096896)
[2021/9/3 下午8:00:05 ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1433761222271963136)
[2021/9/3 下午2:00:15 ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1433670624454119424)
[2021/9/3 上午7:59:43 ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1433580026854379520)
[2021/9/3 上午1:59:22 ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1433489204028903424)


/kind failing-test
<!-- If this is a flake, please add: /kind flake -->

<!-- Please assign a SIG using: /sig SIG-NAME -->
/sig testing

This may be related with  https://github.com/prometheus/client_golang/issues/716 \  https://github.com/prometheus/client_golang/issues/733

--- a/pkg/kubelet/kuberuntime/instrumented_services_test.go
+++ b/pkg/kubelet/kuberuntime/instrumented_services_test.go
@@ -1,9 +1,12 @@
 /*
 Copyright 2016 The Kubernetes Authors.
+
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
+
     http://www.apache.org/licenses/LICENSE-2.0
+
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

--- a/pkg/kubelet/kuberuntime/instrumented_services_test.go
+++ b/pkg/kubelet/kuberuntime/instrumented_services_test.go
@@ -33,6 +33,7 @@ import (
 func TestRecordOperation(t *testing.T) {
 	// Use local registry
 	var registry = compbasemetrics.NewKubeRegistry()
+	defer registry.Reset()
 	registry.MustRegister(metrics.RuntimeOperations)
 	registry.MustRegister(metrics.RuntimeOperationsDuration)
 	registry.MustRegister(metrics.RuntimeOperationsErrors)
@@ -64,8 +65,6 @@ func TestRecordOperation(t *testing.T) {
 	assert.HTTPBodyContains(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		mux.ServeHTTP(w, r)
 	}), "GET", prometheusURL, nil, runtimeOperationsDurationExpected)
-
-	registry.Reset()
 }
 
 func TestInstrumentedVersion(t *testing.T) {

--- a/pkg/kubelet/kuberuntime/instrumented_services_test.go
+++ b/pkg/kubelet/kuberuntime/instrumented_services_test.go
@@ -19,7 +19,6 @@ package kuberuntime
 import (
 	"net"
 	"net/http"
-	"sync"
 	"testing"
 	"time"
 
@@ -30,8 +29,6 @@ import (
 	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
 	"k8s.io/kubernetes/pkg/kubelet/metrics"
 )
-
-var registerMetrics sync.Once
 
 func TestRecordOperation(t *testing.T) {
 	// Use local registry

--- a/pkg/kubelet/kuberuntime/instrumented_services_test.go
+++ b/pkg/kubelet/kuberuntime/instrumented_services_test.go
@@ -22,7 +22,6 @@ import (
 	"testing"
 	"time"
 
-	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"github.com/stretchr/testify/assert"
 
 	compbasemetrics "k8s.io/component-base/metrics"
@@ -33,6 +32,7 @@ import (
 func TestRecordOperation(t *testing.T) {
 	// Use local registry
 	var registry = compbasemetrics.NewKubeRegistry()
+	var gather compbasemetrics.Gatherer = registry
 	defer registry.Reset()
 	registry.MustRegister(metrics.RuntimeOperations)
 	registry.MustRegister(metrics.RuntimeOperationsDuration)
@@ -44,7 +44,7 @@ func TestRecordOperation(t *testing.T) {
 
 	prometheusURL := "http://" + l.Addr().String() + "/metrics"
 	mux := http.NewServeMux()
-	handler := promhttp.HandlerFor(registry, promhttp.HandlerOpts{})
+	handler := compbasemetrics.HandlerFor(gather, compbasemetrics.HandlerOpts{})
 	mux.Handle("/metrics", handler)
 	server := &http.Server{
 		Addr:    l.Addr().String(),

--- a/pkg/kubelet/kuberuntime/instrumented_services_test.go
+++ b/pkg/kubelet/kuberuntime/instrumented_services_test.go
@@ -33,10 +33,12 @@ func TestRecordOperation(t *testing.T) {
 	// Use local registry
 	var registry = compbasemetrics.NewKubeRegistry()
 	var gather compbasemetrics.Gatherer = registry
-	defer registry.Reset()
+
 	registry.MustRegister(metrics.RuntimeOperations)
 	registry.MustRegister(metrics.RuntimeOperationsDuration)
 	registry.MustRegister(metrics.RuntimeOperationsErrors)
+
+	registry.Reset()
 
 	l, err := net.Listen("tcp", "127.0.0.1:0")
 	assert.NoError(t, err)

--- a/pkg/kubelet/kuberuntime/instrumented_services_test.go
+++ b/pkg/kubelet/kuberuntime/instrumented_services_test.go
@@ -1,12 +1,9 @@
 /*
 Copyright 2016 The Kubernetes Authors.
-
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
-
     http://www.apache.org/licenses/LICENSE-2.0
-
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -19,20 +16,27 @@ package kuberuntime
 import (
 	"net"
 	"net/http"
+	"sync"
 	"testing"
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"github.com/stretchr/testify/assert"
 
-	"k8s.io/component-base/metrics/legacyregistry"
+	compbasemetrics "k8s.io/component-base/metrics"
 	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
 	"k8s.io/kubernetes/pkg/kubelet/metrics"
 )
 
+var registerMetrics sync.Once
+
 func TestRecordOperation(t *testing.T) {
-	legacyregistry.MustRegister(metrics.RuntimeOperations)
-	legacyregistry.MustRegister(metrics.RuntimeOperationsDuration)
-	legacyregistry.MustRegister(metrics.RuntimeOperationsErrors)
+	// Use local registry
+	var registry = compbasemetrics.NewKubeRegistry()
+	registry.MustRegister(metrics.RuntimeOperations)
+	registry.MustRegister(metrics.RuntimeOperationsDuration)
+	registry.MustRegister(metrics.RuntimeOperationsErrors)
 
 	l, err := net.Listen("tcp", "127.0.0.1:0")
 	assert.NoError(t, err)
@@ -41,7 +45,8 @@ func TestRecordOperation(t *testing.T) {
 	prometheusURL := "http://" + l.Addr().String() + "/metrics"
 	mux := http.NewServeMux()
 	//lint:ignore SA1019 ignore deprecated warning until we move off of global registries
-	mux.Handle("/metrics", legacyregistry.Handler())
+	handler := promhttp.InstrumentMetricHandler(prometheus.DefaultRegisterer, promhttp.HandlerFor(registry, promhttp.HandlerOpts{}))
+	mux.Handle("/metrics", handler)
 	server := &http.Server{
 		Addr:    l.Addr().String(),
 		Handler: mux,
@@ -61,6 +66,8 @@ func TestRecordOperation(t *testing.T) {
 	assert.HTTPBodyContains(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		mux.ServeHTTP(w, r)
 	}), "GET", prometheusURL, nil, runtimeOperationsDurationExpected)
+
+	registry.Reset()
 }
 
 func TestInstrumentedVersion(t *testing.T) {

--- a/pkg/kubelet/kuberuntime/instrumented_services_test.go
+++ b/pkg/kubelet/kuberuntime/instrumented_services_test.go
@@ -23,7 +23,6 @@ import (
 	"testing"
 	"time"
 
-	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"github.com/stretchr/testify/assert"
 
@@ -48,7 +47,7 @@ func TestRecordOperation(t *testing.T) {
 	prometheusURL := "http://" + l.Addr().String() + "/metrics"
 	mux := http.NewServeMux()
 	//lint:ignore SA1019 ignore deprecated warning until we move off of global registries
-	handler := promhttp.InstrumentMetricHandler(prometheus.DefaultRegisterer, promhttp.HandlerFor(registry, promhttp.HandlerOpts{}))
+	handler := promhttp.HandlerFor(registry, promhttp.HandlerOpts{})
 	mux.Handle("/metrics", handler)
 	server := &http.Server{
 		Addr:    l.Addr().String(),

--- a/pkg/kubelet/kuberuntime/instrumented_services_test.go
+++ b/pkg/kubelet/kuberuntime/instrumented_services_test.go
@@ -43,7 +43,6 @@ func TestRecordOperation(t *testing.T) {
 
 	prometheusURL := "http://" + l.Addr().String() + "/metrics"
 	mux := http.NewServeMux()
-	//lint:ignore SA1019 ignore deprecated warning until we move off of global registries
 	handler := promhttp.HandlerFor(registry, promhttp.HandlerOpts{})
 	mux.Handle("/metrics", handler)
 	server := &http.Server{

--- a/staging/src/k8s.io/component-base/metrics/testutil/metrics_test.go
+++ b/staging/src/k8s.io/component-base/metrics/testutil/metrics_test.go
@@ -587,6 +587,8 @@ func TestGetHistogramVecFromGatherer(t *testing.T) {
 			if diff := cmp.Diff(tt.wantVec, histogramVec); diff != "" {
 				t.Errorf("Got unexpected HistogramVec (-want +got):\n%s", diff)
 			}
+
+			registry.Reset()
 		})
 	}
 }

--- a/staging/src/k8s.io/component-base/metrics/testutil/metrics_test.go
+++ b/staging/src/k8s.io/component-base/metrics/testutil/metrics_test.go
@@ -1,12 +1,9 @@
 /*
 Copyright 2020 The Kubernetes Authors.
-
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
-
     http://www.apache.org/licenses/LICENSE-2.0
-
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -20,13 +17,11 @@ import (
 	"fmt"
 	"math"
 	"reflect"
-	"sync"
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
 	dto "github.com/prometheus/client_model/go"
 	"k8s.io/component-base/metrics"
-	"k8s.io/component-base/metrics/legacyregistry"
 	"k8s.io/utils/pointer"
 )
 
@@ -517,7 +512,6 @@ func TestHistogramVec_Validate(t *testing.T) {
 }
 
 func TestGetHistogramVecFromGatherer(t *testing.T) {
-	var registerMetrics sync.Once
 	tests := []struct {
 		name    string
 		lvMap   map[string]string
@@ -576,16 +570,17 @@ func TestGetHistogramVecFromGatherer(t *testing.T) {
 				Buckets:   buckets,
 			}
 			vec := metrics.NewHistogramVec(HistogramOpts, labels)
-			registerMetrics.Do(func() {
-				legacyregistry.MustRegister(vec)
-			})
+			// Use local registry
+			var registry = metrics.NewKubeRegistry()
+			var gather metrics.Gatherer = registry
+			registry.MustRegister(vec)
 			// Observe two metrics with same value for label1 but different value of label2.
 			vec.WithLabelValues("value1-0", "value2-0").Observe(1.5)
 			vec.WithLabelValues("value1-0", "value2-1").Observe(2.5)
 			vec.WithLabelValues("value1-1", "value2-0").Observe(3.5)
 			vec.WithLabelValues("value1-1", "value2-1").Observe(4.5)
 			metricName := fmt.Sprintf("%s_%s_%s", HistogramOpts.Namespace, HistogramOpts.Subsystem, HistogramOpts.Name)
-			histogramVec, _ := GetHistogramVecFromGatherer(legacyregistry.DefaultGatherer, metricName, tt.lvMap)
+			histogramVec, _ := GetHistogramVecFromGatherer(gather, metricName, tt.lvMap)
 			if diff := cmp.Diff(tt.wantVec, histogramVec); diff != "" {
 				t.Errorf("Got unexpected HistogramVec (-want +got):\n%s", diff)
 			}

--- a/staging/src/k8s.io/component-base/metrics/testutil/metrics_test.go
+++ b/staging/src/k8s.io/component-base/metrics/testutil/metrics_test.go
@@ -1,9 +1,12 @@
 /*
 Copyright 2020 The Kubernetes Authors.
+
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
+
     http://www.apache.org/licenses/LICENSE-2.0
+
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

--- a/staging/src/k8s.io/component-base/metrics/testutil/metrics_test.go
+++ b/staging/src/k8s.io/component-base/metrics/testutil/metrics_test.go
@@ -587,8 +587,6 @@ func TestGetHistogramVecFromGatherer(t *testing.T) {
 			if diff := cmp.Diff(tt.wantVec, histogramVec); diff != "" {
 				t.Errorf("Got unexpected HistogramVec (-want +got):\n%s", diff)
 			}
-
-			registry.Reset()
 		})
 	}
 }

